sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,539 @@
+# Dynamic linking of device code
+
+This document describes purpose and design of dynamic linking of device code
+feature.
+
+## Background
+Sometimes users want to link device code dynamically at run time. One possible
+use case for such linkage - providing device functions via shared libraries.
+The example below shows how device function `LibDeviceFunc` can be dynamically
+linked to a SYCL app:
+```
+// app.cpp
+SYCL_EXTERNAL int LibDeviceFunc(int i);
+class KernelName;
+/* ... */
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for<KernelName>(/* ... */ [=](sycl::item i) {
+  out[i] = LibDeviceFunc(i);
+}); /* ... */
+std::cout << out[i] << “ “;
+
+// lib.cpp
+int SYCL_EXTERNAL LibDeviceFunc(int i) {
+  return i * 2;
+}
+
+// Commands
+clang++ -fsycl lib.cpp -shared -o helpers.so
+clang++ -fsycl app.cpp -lhelpers -o a.out
+./a.out
+Output: 0 2 4 6…
+
+```
+It is possible to manually create `sycl::program` in both app and shared
+library, then use `link` SYCL API to get a single program and launch kernels
+using it. But it is not user-friendly and it is very different from regular
+C/C++ workflow.
+
+Another possible scenario - use functions defined in a pre-compiled device image
+provided by the user. Example:
+```
+// a.cpp
+SYCL_EXTERNAL void LibDeviceFunc();
+...
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for([]() { LibDeviceFunc(); });
+});
+
+// b.cpp
+/*no SYCL_EXTERNAL*/ void LibDeviceFunc() { ... }
+```
+We have a `SYCL_EXTERNAL` function `foo` called from a kernel, but the
+application defined only host version of this function. Then user adds device
+image with definition of `foo` to the fat object via special compiler option
+(like `-fsycl-add-targets`).
+
+The main purpose of this feature is to provide a mechanism which allows to
+link device code dynamically at runtime, such as in the scenarios above.
+
+## Requirements:
+User's device code that consists of some device API (`SYCL_EXTERNAL` functions),
+is compiled into some form and it is not linked statically with device code of
+application. It can be a shared library that contains some device code or a
+separate device image supplied with properties attached. This code is linked
+dynamically at run time with device code of a user's application in order to
+resolve dependencies.
+For this combination the following statements must be true:
+
+The presented dynamic device code linkage mechanism must:
+
+- Allow to represent the actual dynamically linked code as a device binary image
+  which can be:
+  - Embedded into a host shared object by standard SYCL compiler driver
```

Comment:
> Embedded into a host shared object by standard SYCL compiler driver invocation

Can you re-phrase or clarify? Does this mean that the result of the linking can be saved into a host shared object? Which one? Who invokes driver, when and why?

Reply:
>  Does this mean that the result of the linking can be saved into a host shared object?

No, it means that host shared library may have some device binary image embedded (i.e. it is a "fat" shared library) and this embedded device binary image can be linked dynamically with other device binary images (for example embedded into an application's "fat" binary).

>Who invokes driver, when and why?

User, to compile a shared library. Like in this example https://github.com/intel/llvm/pull/3210/files/9f2b787fdf609f1d81e48eb397262a773e4ed094#diff-016c758d0bc2153af2ad9b0add213f00c9ae352ac19f62dea2689b7f8d6e9c55R28

Reply:
OK. Same ask, please make clarify this in the writing.

Reply:
Sure, I've added clarification.